firmware/src/moving_light.c

```diff
@@ -0,0 +1,359 @@
+  return RDMResponder_BuildSetAck(header);
+}
+
+int MovingLightModel_SetPowerState(const RDMHeader *header,
+                                   const uint8_t *param_data) {
+  if (header->param_data_length != sizeof(uint8_t)) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
+  }
+
+  if (param_data[0] > POWER_STATE_NORMAL) {
+    return RDMResponder_BuildNack(header, NR_DATA_OUT_OF_RANGE);
+  }
+  g_moving_light.power_state = param_data[0];
+  return RDMResponder_BuildSetAck(header);
+}
+
+// Public Functions
+// ----------------------------------------------------------------------------
+void MovingLightModel_Initialize(const MovingLightModelSettings *settings) {
+  g_moving_light.device_hours = 0;
+  g_moving_light.lamp_hours = 0;
+  g_moving_light.lamp_strikes = 0;
+  g_moving_light.device_power_cycles = 0;
+  g_moving_light.lamp_strike_time = 0;
+  g_moving_light.lamp_state = LAMP_OFF;
+  g_moving_light.lamp_on_mode = LAMP_ON_MODE_ON;
+  g_moving_light.display_level = 255;
+  g_moving_light.display_invert = false;
+  g_moving_light.power_state = POWER_STATE_NORMAL;
+  g_moving_light.pan_invert = false;
+  g_moving_light.tilt_invert = false;
+  g_moving_light.pan_tilt_swap = false;
+}
+
+static void MovingLightModel_Activate() {
+  g_responder.def = &RESPONDER_DEFINITION;
+  RDMResponder_ResetToFactoryDefaults();
+
+  g_moving_light.pan_invert = false;
+  g_moving_light.tilt_invert = false;
+  g_moving_light.pan_tilt_swap = false;
+}
+
+static void MovingLightModel_Deactivate() {
+}
+
+static int MovingLightModel_Ioctl(ModelIoctl command, uint8_t *data,
+                                  unsigned int length) {
+  switch (command) {
+    case IOCTL_GET_UID:
+      if (length != UID_LENGTH) {
+        return 0;
+      }
+      RDMResponder_GetUID(data);
+      return 1;
+    default:
+      return 0;
+  }
+}
+
+static int MovingLightModel_HandleRequest(const RDMHeader *header,
+                                          const uint8_t *param_data) {
+  if (!RDMUtil_RequiresAction(g_responder.uid, header->dest_uid)) {
+    return RDM_RESPONDER_NO_RESPONSE;
+  }
+
+  uint16_t sub_device = ntohs(header->sub_device);
+
+  // No subdevices.
+  if (sub_device != SUBDEVICE_ROOT && sub_device != SUBDEVICE_ALL) {
+    return RDMResponder_BuildNack(header, NR_SUB_DEVICE_OUT_OF_RANGE);
+  }
+
+  if (header->command_class == DISCOVERY_COMMAND) {
+    return RDMResponder_HandleDiscovery(header, param_data);
+  }
+
+  // This model has no sub devices.
+  if (header->command_class == GET_COMMAND && sub_device == SUBDEVICE_ALL) {
+    return RDMResponder_BuildNack(header, NR_SUB_DEVICE_OUT_OF_RANGE);
+  }
+
+  return RDMResponder_DispatchPID(header, param_data);
+}
+
+static void MovingLightModel_Tasks() {
+  if (g_moving_light.lamp_state == LAMP_STRIKE &&
+      CoarseTimer_HasElapsed(g_moving_light.lamp_strike_time,
+                             LAMP_STRIKE_DELAY)) {
+    g_moving_light.lamp_state = LAMP_ON;
+    g_moving_light.lamp_strikes++;
```

Comment:
Ah, I see what you're saying, I think you should only be able to set one of the states though. In fact this is probably a bug in OLA too, you either want to strike the lamp (the lamp on command normally in desks), or turn it off. Meanwhile it may be in a number of additional states, e.g. striking, not present, standby. If you could just turn it on, no-one would bother striking it.